---
title: "HW2"
author: "slendersold"
date: "2024-04-20"
output: html_document
---
# Data preprocessing and normalization
Let's read peaks_data.csv to df variable. We will navigate through the whole data set by Gene_id field. This field have nan values and duplicates. We can omit these rows or aggregate them in case of duplicated value. Also let's get rid of first 4 columns which brings us nothing but burden. So this is what we get:

```{r setup, include=FALSE}
df = read.csv('C:/Users/slendersold/Documents/Учеба/Мага Биоинф/proteomix/peaks_data.csv')  
df = df[!df$Gene_id %in% c("nan",""),] 
df$Gene_id[which(duplicated(df$Gene_id))] 
df = aggregate(df, by = list(df$Gene_id), mean) 
df = df[!grepl("',",df$Group.1),] 
rownames(df) = df$Group.1 
df = df[-c(1,2,3,4)]
head(df)
```

Lets filter uninformative rows with humber of NA values is more than a half

```{r}
genes_with_NA <- names(which(rowSums(is.na(df)) > ncol(df)/2))
df <- df[!rownames(df) %in% genes_with_NA,]
```
```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(impute)
```

Lets use impute.knn() function to impute missing expression data, using nearest neighbor averaging

```{r}
df <- as.matrix(df)
df_trans <- t(df)
knn_df <- impute.knn(df_trans, k = 5)
knn_df <- knn_df$data
knn_df <- as.data.frame(knn_df)
knn_df <- t(as.matrix(knn_df))
```
```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(RColorBrewer)
```

Lets draw boxplot of the initial data set

```{r, warning=FALSE, message=FALSE, echo=FALSE}
experiment <- as.factor(c("1","1","1","1","1","1","1","1","1","1","1","1","1","1","2","2","2","2","2","2","2","2","2","2","2","2","2","2","2","2","2","2","2"))
pal <- brewer.pal(n = 9, name = "Set1")
cols <- pal[experiment]
boxplot(knn_df, outline = FALSE, main = "initial data", col = cols)
legend("topright", levels(experiment), fill = pal, bty = "n", xpd = T)
```
```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(limma)
```

After initial clearing of data we can perform normalization

```{r}
df_log <- log2(knn_df+1)
df_norm <- normalizeQuantiles(as.matrix(df_log))
```
```{r, warning=FALSE, message=FALSE, echo=FALSE}
boxplot(df_norm, outline = FALSE, main = "normalized data", col = cols)
legend("topright", levels(experiment), fill = pal, bty = "n", xpd = T)
```

# Evaluation of the presence of a batch-effect

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(vegan)
library(ggplot2)
```

Lets perform PCA plot to look at clusters

```{r}
pca <- t(df_norm)
terapod_pca <- rda(pca, scale = TRUE)

rownames(pca) <- c("1","1","1","1","1","1","1","1","1","1","1","1","1","1","2","2","2","2","2","2","2","2","2","2","2","2","2","2","2","2","2","2","2")

df_scores <- data.frame(pca,
                        scores(terapod_pca, display = "sites", choices = c(1, 2, 3), scaling = "sites"))
```
```{r, warning=FALSE, message=FALSE, echo=FALSE}
p_scores <- ggplot(df_scores, aes(x = PC1, y = PC2)) + 
  geom_point(aes(color = rownames(pca)), alpha = 0.5) +
  coord_equal(xlim = c(-3, 3), ylim = c(-3, 3)) + ggtitle(label = "PCA") + theme_bw()
p_scores
```

As we can see clusters looks as usual

```{r, warning=FALSE, message=FALSE, echo=FALSE}
maplot <- function(X1, X2, pch = 21, main = "MA-plot", xlab = "Average log-expression", ylab = "Expression log-ratio", lpars = list(col = "blue", lwd = 2), ...){
  # Координаты
  X <- (rowMeans(X2) + rowMeans(X1)) / 2
  Y <- rowMeans(X2) - rowMeans(X1)
  # График
  scatter.smooth(x = X, y = Y, main = main, pch = pch, xlab = xlab, ylab = ylab, lpars = lpars, ...)
  abline(h = c(-1, 0, 1), lty = c(2, 1, 2))
}

```

Lets draw maplot

```{r, warning=FALSE, message=FALSE, echo=FALSE}
maplot(df_norm[,1:14], df_norm[,15:33])
```

Line does not overshoot middle zone so maplot does not show the presence of batch effects

# Identification of differentially expressed proteins

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(Biobase)
```

Lets perform differential expression 

```{r}
# Данные экспрессии
expr_data <- as.matrix(df_norm)

# Данные о пробах
pheno_data <- data.frame(experiment)
rownames(pheno_data) <- colnames(df_norm)
pheno_metadata <- data.frame(
  labelDescription = c("Experimental condition"),
  row.names=c("Condition"))
pheno_data <- new("AnnotatedDataFrame",
                 data = pheno_data,
                 varMetadata = pheno_metadata)

# Данные о признаках (белках)
feature_data <- data.frame(Prot = rownames(expr_data))
rownames(feature_data) <- rownames(expr_data)
feature_metadata <- data.frame(
  labelDescription = c("Protain name"),
  row.names = c("Protain"))
f_data <- new("AnnotatedDataFrame",
              data = feature_data,
              varMetadata = feature_metadata)

# Собираем вместе
exp_set <-
  ExpressionSet(assayData = expr_data,
                phenoData = pheno_data,
                featureData = f_data)
```

fit linear model for each gene given a series of arrays, compute moderated t-statistics, moderated F-statistic, and log-odds of differential expression by empirical Bayes moderation of the standard errors towards a global value

```{r}
X <- model.matrix(~ experiment, pData(exp_set))
fit <- lmFit(exp_set, design = X, method = "robust", maxit = 1000)
efit <- eBayes(fit)
```

draw new MA plot

```{r, warning=FALSE, message=FALSE, echo=FALSE}
MA_limma <- function(efit, coef, n = 10, signif = TRUE, fdr = 0.05, lfc = 0, text = TRUE, cex.text = 0.8, col.text = "grey20", main = "MA-plot", xlab = "Average log-expression", ylab = "Expression log-ratio", pch = 19, pch.signif = 21, col = "darkgreen", alpha = 0.3, cex = 0.3, ...){
  # соотношение и интенсивность
  R <- efit$coefficients[, coef]
  I <- efit$Amean
  # прозрачный цвет
  col_btransp <- adjustcolor(col, alpha.f = alpha)
  # график
  plot(I, R, cex = cex, main = main, pch = pch, xlab = xlab, ylab = ylab, col = col_btransp, ...)
  abline(h = 0)
  # отмечаем дифференциально-экспрессируемые белки
  if(signif){
    sign <- p.adjust(efit$p.value[, coef], method = "BH") <= fdr
    large <- abs(efit$coefficients[, coef]) >= lfc
    points(I[sign & large], R[sign & large], cex = cex*2, col = "orange2", pch = pch.signif)
  }
  # подписываем первые n белков с сильнее всего различающейся экспрессией
  if(text){
    ord <- order(efit$lods[, coef], decreasing = TRUE)
    top_n <- ord[1:n]
    text(I[top_n], R[top_n], labels = efit$genes[top_n, ], pos = 4, cex = cex.text, col = col.text)
  }
}


MA_limma(efit, coef = 2, n = 30)
```

and show up and down regulated genes

```{r}
# ## Добываем дифференциально-экспрессируемые белки для дальнейшей работы =======
# Первые 20 дифференциальных белков
my_list <- topTable(efit, coef = 2, n = 100)
# Фильтруем ExpressionSet
dif_exp_set <- exp_set[fData(exp_set)$Prot %in% my_list$Prot, ]
```
```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(gplots)
```

Heatmap

```{r fig.height=10, fig.width=15, warning=FALSE, message=FALSE, echo=FALSE}
dat <- as.matrix(exprs(dif_exp_set))
# после дополнительной стандартизации по белкам
pal_blue_red <- colorpanel(75, low = "steelblue", mid = "white", high = "red")
heatmap.2(dat, col = pal_blue_red, scale = "row", key = TRUE, symkey = FALSE, density.info = "none", trace = "none", cexRow = 0.9, cexCol = 1, margins = c(4, 3), keysize = 0.8, key.par = list(mar = c(3, 0.1, 3, 0.1)))
```

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(EnhancedVolcano)
```

EnhancedVolcano plot

```{r Volcano, warning=FALSE, message=FALSE, fig.height=10, fig.width=15}
topTable(efit, coef = 2)
numGenes <- nrow(exprs(exp_set))
res = topTable(efit, number = numGenes)

EnhancedVolcano(res,
                lab = rownames(res),
                x = 'logFC',
                y = 'adj.P.Val',
                title = 'CJK versus BT',
                FCcutoff = 2,
                pointSize = 3.0,
                labSize = 6.0)
```

# Enrichment
```{r GO, warning=FALSE, message=FALSE, echo=FALSE}
library(org.Hs.eg.db)
library(clusterProfiler)
```

## Up-regulated genes
```{r GOup, warning=FALSE, message=FALSE, fig.height=10, fig.width=15}
genes_up = res$Prot[(res$logFC>2 & res$adj.P.Val<0.05)]

GO_enrich <- enrichGO(genes_up, "org.Hs.eg.db", keyType = "SYMBOL", ont = "BP")
```

```{r, warning=FALSE, message=FALSE, echo=FALSE}
head(GO_enrich@result)
```

```{r, warning=FALSE, message=FALSE, echo=FALSE}
dotplot(GO_enrich, showCategory = 20)
```

```{r, warning=FALSE, message=FALSE, echo=FALSE}
GO_enrich <- enrichplot::pairwise_termsim(GO_enrich)
```

```{r, warning=FALSE, message=FALSE, echo=FALSE}
emapplot(GO_enrich)
```

```{r, warning=FALSE, message=FALSE, echo=FALSE}
goplot(GO_enrich)
```

## Down-regulated genes
```{r GOdown, warning=FALSE, message=FALSE, fig.height=10, fig.width=15}
genes_down = res$Prot[(res$logFC<2 & res$adj.P.Val<0.05)]

GO_enrich <- enrichGO(genes_down, "org.Hs.eg.db", keyType = "SYMBOL", ont = "BP")
```

```{r, warning=FALSE, message=FALSE, echo=FALSE}
head(GO_enrich@result)
```

```{r, warning=FALSE, message=FALSE, echo=FALSE}
dotplot(GO_enrich, showCategory = 20)
```

```{r, warning=FALSE, message=FALSE, echo=FALSE}
GO_enrich <- enrichplot::pairwise_termsim(GO_enrich)
```

```{r, warning=FALSE, message=FALSE, echo=FALSE}
emapplot(GO_enrich)
```

```{r, warning=FALSE, message=FALSE, echo=FALSE}
goplot(GO_enrich)
```

Тhe GO-enrichment analysis suggests that the myocardium in CJK patients is facing challenges related to toxic substance response, doxorubicin metabolism, and disruptions in essential cellular processes. These molecular disturbances could be contributing factors to the cardiac dysfunction and structural changes observed in these patients.